src/validate.js

```diff
@@ -159,6 +160,368 @@ function transformAjvErrors(errors = []) {
   });
 }
 
+/**
+ * Iterates over all {_errors} to create an object containing properties with the
+ * error property as the key and all errors for this error property as the value
+ *
+ * @example
+ * // sample ajv _errors input
+ * [
+ *    {property: '.foo', ...errorForThisProp},
+ *    {property: '.bar', ...errorForThisProp},
+ *    {property: '.bar[2]', ...errorForThisProp},
+ *    {property: '.bar[2].qux', ...errorForThisProp},
+ *    {property: '.bar', ...errorForThisProp}
+ *    {property: '', ...errorForThisProp}
+ * ]
+ * // sample output
+ * // different cases possible of an error.properties; here are some EXAMPLES
+ * {
+ *     ".foo": [* all errors for this prop *],
+ *     ".bar[n]": [* all errors for this prop *],
+ *     ".bar[n].qux": [* all errors for this prop *],
+ *     ".bar[n].qux[4]": [* all errors for this prop *],
+ *     ".bar[n].qux[4].wibble": [* all errors for this prop *],
+ *     ".bax.2": [* all errors for this prop *],
+ *     "": [* all errors for this prop *]
+ * }
+ *
+ * @param {object[]} _errors - the errors of ajv
+ */
+function accumulatePropertyErrors(_errors) {
+  const flattenedErrorSchema = {};
+  const allErrorKeysWithoutDuplicates = [
+    ...new Set(_errors.map(err => err.property)),
+  ];
+  allErrorKeysWithoutDuplicates.forEach(errKey => {
+    const errPropKey =
+      errKey && errKey.startsWith(".") ? errKey.substring(1) : errKey;
+    flattenedErrorSchema[errPropKey] = _errors.filter(
+      err => err.property === errKey
+    );
+  });
+  return flattenedErrorSchema;
+}
+
+function removeDotPrefixFromString(string) {
+  return string.startsWith(".") ? string.substring(1) : string;
+}
+
+function addErrorToArrayIfAbsent(array) {
```

Comment:
Hey @Gurkenschreck , thanks for the change. I haven't gotten a chance to step through all your logic completely yet, but given that you've added around 12 different helper functions, I want to make sure that when adding this much complexity, we offset this complexity by adding unit tests. You do have tests for the uiSchema ordering -- which is great -- but could you also look into adding unit tests for some of the more important helper functions? That will help a ton, both in reviewing the code to see what each function is supposed to do, and in also changing / maintaining this logic in the future.

Reply:
Hi,
I see your point and yes, it would help. But I'd want to avoid exporting those functions, making the helper functions public, just for the sake of testing (for the same reasons as one would not make every method public in e.g. Java just to test it).
Would it be enough to think about more test cases that I can add to test the (now renamed) `sortErrorsByUiSchema` function?
If not, I don't wanna press this paradigm on this project, so I can export those functions and test them. What do you think?

Thanks for the feedback!

Reply:
I see your point, but I think it's ok to export them because we do want to make sure each helper function is working as intended. Also, it's not super clear to me what functions like `filterErrorsByUnmentionedUiorder` and `recursivelyAddErrorsToArrayForNode` exactly do (unless I step through the code exactly), so adding maybe 1-2 tests for each of these functions would help to document exactly what each of these functions is supposed to do.

The tests may not need to be exhaustive, given that you already have many tests for `sortErrorsByUiSchema`.